### PR TITLE
EZP-29328: Allow docbook tag "programlisting"

### PR DESCRIPTION
**Pull request by vidarl**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29328](https://jira.ez.no/browse/EZP-29328)
| **Bug/Improvement**| improvment
| **New feature**    | yes
| **Target version** | `6.13`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

We need `<programlisting>` for literal, see [EZP-29328](https://jira.ez.no/browse/EZP-29328)
See corresponding [PR for conversion script ( ezxmltext -> docbook )](https://github.com/ezsystems/ezplatform-xmltext-fieldtype/pull/71)

With this PR, the alloy editor is able to edit existing programlistings, but button for creating new programlisting should be created in a separate admin-ui PR.

**TODO**:
- [x] Move to 6.13
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by vidarl** [comment]
FYI : Added initial support for  converting `<programlisting>` tag in docbook->xhtml (output) transformation

(edit) not supported
Neither is `class="html"` which requires additional support in custom tags

Reserved html characters should be encoded as html entities on save to db ( or we have to deal with that during XSLT transformation, which will be problematic

File: eZ/Publish/Core/FieldType/RichText/Resources/schemas/ezxhtml5/output/ezxhtml5.xsd [hunk 1]
```diff
@@ -249,13 +249,21 @@
     <xs:attributeGroup ref="globalAttributeGroup"/>
   </xs:complexType>
 
+  <xs:complexType name="prePhrasingContentElement">
```
Comment (vidarl) [inline eZ/Publish/Core/FieldType/RichText/Resources/schemas/ezxhtml5/output/ezxhtml5.xsd]: AFAIK, it is not possible to add generic support for data-* attributes i xsd schemas. ( no such wildcard functionality exists)

Comment (vidarl) [reply eZ/Publish/Core/FieldType/RichText/Resources/schemas/ezxhtml5/output/ezxhtml5.xsd]: The point with `data-language` is a request from @andrerom.
In legacy, you may add a custom tag named `language` to the `<literal>` tag. The idea is to store the value of such a custom attribute (if it is defined) in `data-language`.

But it then might be applicable to add a template rendering step for `<programlisting>`  so user may deal with  that attribute in a template?

**Comment by vidarl** [comment]
> How is this interpreted by Alloy ?

It's fair to say that it is **not** interpreted very well... So work on that part is needed by someone who knows that stuff.
"emulating" a custom tag might be a possibility

**Comment by andrerom** [comment]
@bdunogier suggestion from my side was to map this top <pre>, which alloy supports out of the box as a Style, see their demo on front page (it's not mentioned in their doc):

screenshot 2018-09-27 at 14 09 41

**Comment by vidarl** [comment]
´<pre>´ seems to work fine with alloy, indeed. I'll add that to the PR

**Comment by vidarl** [comment]
transformed to `<pre>` for Alloy and rebased on 6.13.

Should be ready for final review now

File: eZ/Publish/Core/FieldType/Tests/RichText/Converter/Xslt/_fixtures/docbook/033-programlisting.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<section xmlns="http://docbook.org/ns/docbook" xmlns:ezcustom="http://ez.no/xmlns/ezpublish/docbook/custom" xmlns:ezxhtml="http://ez.no/xmlns/ezpublish/docbook/xhtml" xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0-variant ezpublish-1.0">
+    <programlisting><![CDATA[Literal with CNAME end sequence here : ]]]]><![CDATA[> and some more text here]]></programlisting>
+    <programlisting><![CDATA[ordinary literal]]></programlisting>
+    <programlisting><![CDATA[more
+            complicated
+             literal
+              tag]]></programlisting>
+    <para>the php one:</para>
+    <programlisting ezxhtml:class="php"><![CDATA[function foobar()
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Tests/RichText/Converter/Xslt/_fixtures/docbook/033-programlisting.xml]: At the first glance it looks like a typo and supposedly should be `language`. I know that it doesn't matter, but I suggest using something similar to what you're using below, so e.g. `blue`. Let's think of first-time, inexperienced readers :)

Comment (vidarl) [reply eZ/Publish/Core/FieldType/Tests/RichText/Converter/Xslt/_fixtures/docbook/033-programlisting.xml]: The reason why I wrote it like this is that those who haven't created a custom attribute "language" in legacy, likely would have used the class for this purpose instead.
But I'll change it to blue as you suggested

File: eZ/Publish/Core/FieldType/RichText/Converter/CDataInProgramListing.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\FieldType\RichText\Converter;
+
+use eZ\Publish\Core\FieldType\RichText\Converter;
+use DOMDocument;
+use DOMXPath;
+
+class CDataInProgramListing implements Converter
```
[the remark this answers is not in the document]
Comment (vidarl) [reply eZ/Publish/Core/FieldType/RichText/Converter/CDataInProgramListing.php]: Fixed as suggested

File: eZ/Publish/Core/FieldType/Tests/RichText/Converter/Xslt/_fixtures/docbook/033-programlisting.xml (hunk 2 above)
Comment (vidarl) [reply eZ/Publish/Core/FieldType/Tests/RichText/Converter/Xslt/_fixtures/docbook/033-programlisting.xml]: Fixed

**Comment by vidarl** [comment]
@andrerom , @alongosz : Fixed review comments and rebased

**Pull request merged**